src/framework/entity.js

```diff
@@ -337,15 +347,63 @@ pc.extend(pc, function () {
 
         var i;
         for (i = 0; i < this._children.length; i++) {
-            var child = this._children[i];
-            if (child instanceof pc.Entity) {
-                c.addChild(child.clone());
+            var oldChild = this._children[i];
+            if (oldChild instanceof pc.Entity) {
+                var newChild = oldChild._cloneRecursively(duplicatedIdsMap);
+                c.addChild(newChild);
+                duplicatedIdsMap[oldChild.getGuid()] = newChild.getGuid();
             }
         }
 
         return c;
     };
 
+    // When an entity that has properties that contain references to other
+    // entities within its subtree is duplicated, the expectation of the
+    // user is likely that those properties will be updated to point to
+    // the corresponding entities within the newly-created duplicate subtree.
+    //
+    // To handle this, we need to search for properties that refer to entities
+    // within the old duplicated structure, find their newly-cloned partners
+    // within the new structure, and update the references accordingly. This
+    // function implements that requirement.
+    function resolveDuplicatedEntityReferenceProperties(oldSubtreeRoot, oldEntity, newEntity, duplicatedIdsMap) {
+        // TODO Would be nice to also make this work for entity script attributes
```

Comment:
I have a sketch of this working locally – we have everything we need to do it, but it needs a bit of cleaning up. Also it's technically a breaking change (if anyone's somehow relying on the old behaviour, although you would think not), so we should treat it with a bit more caution. I'll put it in a separate PR.